datalad_next/iter_collections/gitstatus.py

```diff
@@ -0,0 +1,521 @@
+                        item.add_modification_type(
+                            GitContainerModificationType.untracked_content)
+                else:
+                    # this directory is gone entirely
+                    item.status = GitDiffStatus.deletion
+                    item.modification_types = None
+            # we dealt with this item completely
+            dir_items.pop(item.name, None)
+        if item.status:
+            yield item
+
+    if untracked == 'no':
+        return
+
+    # yield anything untracked, and inspect remaining containers
+    for dir_item in dir_items.values():
+        if dir_item.gitsha is None and dir_item.gittype is None:
+            # this is untracked
+            yield GitDiffItem(
+                # for homgeneity for report a str-path no matter what
+                name=str(dir_item.name),
+                status=GitDiffStatus.other,
+            )
+        elif dir_item.gittype in container_types:
+            # none of these containers has any modification other than
+            # possibly untracked content
+            item = GitDiffItem(
+                # for homgeneity for report a str-path no matter what
+                name=str(dir_item.name),
+                # this submodule has not been detected as modified
+                # per-commit, assign reported gitsha to pre and post
+                # state
+                gitsha=dir_item.gitsha,
+                prev_gitsha=dir_item.gitsha,
+                gittype=dir_item.gittype,
+                # TODO others?
+            )
+            if item.gittype == GitTreeItemType.submodule:
+                # issue standard submodule container report
+                _eval_submodule(path, item, eval_submodule_state)
+            else:
+                # this is on a directory. if it appears here, it has
+                # no modified content
+                if _path_has_untracked(path / dir_item.path):
+                    item.status = GitDiffStatus.modification
+                    item.add_modification_type(
+                        GitContainerModificationType.untracked_content)
+            if item.status:
+                yield item
+
+
+def _yield_repo_items(
+    path: Path,
+    untracked: str | None,
+    eval_submodule_state: str,
+    present_submodules: dict[str, GitTreeItem],
+) -> Generator[GitDiffItem, None, None]:
+    """Report status items for a single/whole repsoitory"""
+    # start with a repository-contrained diff against the worktree
+    for item in iter_gitdiff(
+        path,
+        from_treeish='HEAD',
+        # to the worktree
+        to_treeish=None,
+        recursive='repository',
+        # TODO check if we can go cheaper. We need to redo some check for
+        # adjusted mode, and other cases anyways
+        eval_submodule_state='commit'
+        if eval_submodule_state == 'full' else eval_submodule_state,
+        #eval_submodule_state=eval_submodule_state,
+    ):
```

Comment:
I found an edge case: when `datalad next-status` is called in an empty git repo, this happens:

```
❱ git init
Initialized empty Git repository in /tmp/foo/.git/
❱ datalad next-status
CommandError: 'git diff-index --no-rename-empty --relative --raw -z --ignore-submodules=dirty HEAD' failed with exitcode 128 under /tmp/foo
fatal: ambiguous argument 'HEAD': unknown revision or path not in the working tree.
```

There are no commits, so HEAD points to a `refs/heads/main` which does not exist yet. Both `git status` and `datalad status` work:

```
❱ git status
On branch main

No commits yet

nothing to commit (create/copy files and use "git add" to track)

❱ datalad status
nothing to save, working tree clean
```

(same applies to other `_yield*` functions)

That being said, it won't be relevant to DataLad datasets, because AFAIK `datalad create` always commits.

Reply:
Thanks! I will fix this!

Reply:
This was a bit more complex than anticipated, but it should work now.

I took the opportunity to restructure the implementations a bit and started to push towards #597